### ISSUE TITLE
Split some cmdlets dependent on Windows Compatibility Pack out as separate module

### ISSUE THREAD

**Issue by SteveL-MSFT** [PowerShell/PowerShell#5603]
Some modules that don't exist in PSCore6 should be split into a separate "Windows PowerShell Compatibility Pack" that is dependent on the DotNetCore Windows Compatibility Pack as well WindowsPSModulePath:

- [ ] WMI cmdlets
- [ ] EventLog cmdlets
- [ ] PerfCounter cmdlets

The "Windows PowerShell Compatibility Pack" Module would reside in a separate repo.

**Comment by markekraus** [comment]
Would this separate module contain all of the libraries of the Windows Compatibility Pack? (replacing my PSCoreWindowsCompat module would be ideal). The largest complaint and grumble I hear from the community is about missing .NET types.

**Comment by SteveL-MSFT** [comment]
@markekraus yes. that would be the intent.  I created https://github.com/PowerShell/WindowsPowerShellCompatibilityPack in case you're bored over the holidays :)

Otherwise, I might start refactoring the cmdlets from this repo to that repo over the holidays.

**Comment by markekraus** [comment]
Hmm.. I'll help out a bit on that, but I had planned to use the holidays to address web cmdlets backlog. 😆 . I don't know where to start with it though. Feel free to add me as a collaborator on that repo and I will pitch in with what I can.

**Comment by RichardSiddaway** [comment]
Do we really need the WMI cmdlets in v6.  The CIM cmdlets are already available in v6 and are the recommended way to access CIM/WMI. The v5.1 help points users to the CIM cmdlets rather than the WMI cmdlets.

**Comment by markekraus** [comment]
We don't, no.  But there are a bunch of modules out there that do (for better or worse). All of those fall into the "nice to have" box. Which is why it makes sense to move them out of the core project and into a separate module.

**Comment by iSazonov** [comment]
I wonder if we would implement missing accelerators there ([WMI], [DirectoryService]).

**Comment by SteveL-MSFT** [comment]
I tried pulling all the necessary cmdlet source files to the new repo.  Unfortunately because they are part of Microsoft.PowerShell.Management module, it makes all kinds of assumptions and dependencies on internal code.

**Comment by markekraus** [comment]
Yea. I wouldn't expect that to be easy. Do we have a good list of the cmdlets?

**Comment by iSazonov** [comment]
I tried the same with HelpSystem - it dependencies on internal code too.
I guess we'll save a lot of time if we displace the compatibility module inside the repo as separate assembly.

**Comment by SteveL-MSFT** [comment]
Looks like System.Management isn't part of the dotnet Windows Compatibility Pack and we really need that for the old WMI apis.  So this effort can't start until that shows up which is in the dotnet WIndows Compatibility Pack roadmap.

**Comment by SteveL-MSFT** [comment]
@iSazonov according to their [blog post](https://blogs.msdn.microsoft.com/dotnet/2017/11/16/announcing-the-windows-compatibility-pack-for-net-core/), it's `Coming`

**Comment by markekraus** [comment]
That blog posts also lists `System.DirectoryServices` as `coming` but it is clearly in the package.... That blog post is confusing. unless `coming` means `not completely here`?

**Comment by iSazonov** [comment]
I tried to install the package and see both `System.DirectoryServices` and `System.Management`:
```
log  : Installing System.Buffers 4.5.0-preview1-25914-04.
log  : Installing System.Text.Encoding.CodePages 4.5.0-preview1-25914-04.
log  : Installing System.Security.Permissions 4.5.0-preview1-25914-04.
log  : Installing System.ServiceProcess.ServiceController 4.5.0-preview1-25914-04.
log  : Installing System.Security.Cryptography.ProtectedData 4.5.0-preview1-25914-04.
log  : Installing System.ServiceModel.Syndication 4.5.0-preview1-25914-04.
log  : Installing System.Security.Principal.Windows 4.5.0-preview1-25914-04.
log  : Installing System.Threading.AccessControl 4.5.0-preview1-25914-04.
log  : Installing Microsoft.Win32.Registry 4.5.0-preview1-25914-04.
log  : Installing System.Security.Cryptography.Xml 4.5.0-preview1-25914-04.
log  : Installing System.Security.AccessControl 4.5.0-preview1-25914-04.
log  : Installing System.Security.Cryptography.Pkcs 4.5.0-preview1-25914-04.
log  : Installing System.IO.Pipes.AccessControl 4.5.0-preview1-25914-04.
log  : Installing System.Security.Cryptography.Cng 4.4.0.
log  : Installing System.IO.Ports 4.5.0-preview1-25914-04.
log  : Installing System.Runtime.Caching 4.5.0-preview1-25914-04.
log  : Installing Microsoft.Windows.Compatibility 2.0.0-preview1-25914-04.
log  : Installing Microsoft.Win32.Registry.AccessControl 4.5.0-preview1-25914-04.
log  : Installing Microsoft.Windows.Compatibility.Shims 2.0.0-preview1-25914-04.
log  : Installing System.CodeDom 4.5.0-preview1-25914-04.
log  : Installing System.Configuration.ConfigurationManager 4.5.0-preview1-25914-04.
log  : Installing System.Data.Odbc 4.5.0-preview1-25914-04.
log  : Installing System.Diagnostics.EventLog 4.5.0-preview1-25914-04.
log  : Installing System.Diagnostics.PerformanceCounter 4.5.0-preview1-25914-04.
log  : Installing System.DirectoryServices 4.5.0-preview1-25914-04.
log  : Installing System.DirectoryServices.AccountManagement 4.5.0-preview1-25914-04.
log  : Installing System.DirectoryServices.Protocols 4.5.0-preview1-25914-04.
log  : Installing System.Drawing.Common 4.5.0-preview1-25914-04.
log  : Installing System.IO.FileSystem.AccessControl 4.5.0-preview1-25914-04.
log  : Installing System.IO.Packaging 4.5.0-preview1-25914-04.
log  : Installing System.Management 4.5.0-preview1-25914-04.
```

**Comment by iSazonov** [comment]
What does PowerShell team think about this recommendation from the blog article:
>As you can see in the previous table, about half of the components in the Windows Compatibility Pack are Windows-only, the other half works on any platform. Your code can always assume that all the APIs exist across all platforms, but if they are Windows-only they throw PlatformNotSupportedException. This allows you to write code that calls Windows-only APIs after doing a platform check at runtime, rather than having to use conditional compilation using #if. We recommend you to use RuntimeInformation.IsOSPlatform() for platform checks

**Comment by markekraus** [comment]
@iSazonov I believe this module was targeted for windows-only. The `PSCoreWindowsCompat` module I put out is also targeted that way. It's purpose is to enable the missing features on windows, not to make a cross-plat experience.

**Comment by iSazonov** [comment]
@markekraus I don't understand your comment. The blog post describe recommended porting process and using the module.

**Comment by markekraus** [comment]
@iSazonov I would rather have a separate module that contained the x-plat capable libraries or add them to the core project itself. Mixing windows only an x-plat is bad design, IMO. It's too confusing for the user.

**Comment by iSazonov** [comment]
It's not design. This is the process of porting that is offered by the .Net team. We use preprocessor directives. They're offering to use this Windows Compatibility package, runtime checks, and API analyzer. It looks very productive and we could use that process.

**Comment by kramlat** [comment]
is it possible to have these wrap systemd and other linux tools on linux when the target computer runs linux or what ever the mac users are using if the target computer is a mac?

**Comment by SteveL-MSFT** [comment]
@kramlat I think wrapping some of these so that the cmdlet is an abstraction over the OS is something we'd want to do long term particularly for things like logging, services, computer info.  However, I think we would design those as new cmdlets.

**Comment by iSazonov** [comment]
@kramlat We want our cmdlets to work on _all systems_. The difficulty is that different Unix distributions have many differences.. Specifically, not all Unix systems use system, utilities have different command line parameters, slightly different behavior or output.

**Comment by kramlat** [comment]
Make a conditional check for a OS specific symbol when the module is used and have the code execute relative to what OS it is on, oh and have powershell on linux check for each init system, i.e.sysvinit and systemd for an example.  oh and on bsd systems, bsdinit.

**Comment by iSazonov** [comment]
@kramlat Check the OS pointless - we need to check _features_. But features also have many modifications either in the distributions or can be made by the users.

**Comment by kramlat** [comment]
you have a point about features but I feel both are important but as for user specific changes, those the user made and the source tree does not need to account for those changes, the user since he or she knows what was done, can make the appropriate changes.

**Comment by SteveL-MSFT** [comment]
This work will be done in the new repo linked above

**Comment by iSazonov** [comment]
CoreFX implements more and more APIs. For example WMI already is seems there. Will we move the code back when CoreFX implement needed APIs? This suggests that it is easier and faster to use Windows Compatibility Pack in this repo than to transfer the code to another repo. It will also help us to get rid of `if Unix` and perhaps resolve packaging problems.

**Comment by SteveL-MSFT** [comment]
@iSazonov Our current thinking is that the Windows Compatibility Pack is something users would opt-in and not part of PSCore6.x by default.  Even for WMI, we want users to move to the CIM cmdlets which are more interoperable with 3rd party CIM/WSMan implementations while the WMI cmdlets only work with WMI.

**Comment by iSazonov** [comment]
@SteveL-MSFT Thanks!
I still haven't full understanding. Windows PowerShell Compatibility Pack is mostly for _backward compatibility_ with Windows PowerShell. It only makes sense for those features that will _never_ be ported. The rest needs to be implemented _here_, so far we can do it in part using WCP.

Apart from that we have and will have a lot of places in the code under `if Unix`. We have a lot of issues about the lack of dlls. We already have request to support framework-based distribution (not self-contained #6058). We have problems with Unix packaging. Using WCP and the _methods_ from the aforementioned blog brings us closer to addressing these issues significantly.

**Comment by SteveL-MSFT** [comment]
@iSazonov you are correct that current thinking is that WPCP is primarily for transition from Windows PS to PSCore.  It's not clear to me the future of WCP for dotnetcore and whether we would depend on it beyond Windows PS compatibility.

**Comment by iSazonov** [comment]
@SteveL-MSFT Yes, currently WCP is still only a preview. This is not the main. My understanding is that  WCP is just _meta-package_ and we could use, ex., `System.Directory` directly with using the recommendations fro .Net Core team:
>As you can see in the previous table, about half of the components in the Windows Compatibility Pack are Windows-only, the other half works on any platform. **Your code can always assume that all the APIs exist across all platforms**, but if they are Windows-only they throw PlatformNotSupportedException. This allows you to write code that calls Windows-only APIs after doing a platform check at runtime, rather than having to use conditional compilation using #if. We recommend you to use RuntimeInformation.IsOSPlatform() for platform checks.

My understanding is that this technique allows us:
- to bring back to Windows many features that are now disabled
- exclude conditional compilation using #if UNIX
- exclude dll errors
- even possible to compile without specific RID and get unify packages.

**Comment by markekraus** [comment]
Personally.. I really don't want anything in WCP that is x-plat. I want it targeted for Windows only and to scream bloody murder by throwing 5 page long error messages when you try to use it on macOS or Linux. Anything x-plat should be in a different module or rolled into the Core project.  The reason being is user support. if suddenly WMI cmdlets are visible on linux but not usable and will never be usable... that will result in a constant stream of  "Can't use $WMICmdlet cmdlets on $NotWindowsPlatform" issues and constant complaining of how we are failing to make these never-linux APIs work on linux.

**Comment by iSazonov** [comment]
The idea is to move everything we don't want to port into Windows PowerShell Compatibility module, the rest should stay here and appear in Unix as APIs is ported. So you won't see WMI cmdlets on Unix. For not ported cmdlets we have some options - return user friendly error or don't expose cmdlet at all (as we are now doing).

**Comment by iSazonov** [comment]
By the way it will ease PSScriptAnalyzer getting the information ported cmdlet (or parameter) or not.
